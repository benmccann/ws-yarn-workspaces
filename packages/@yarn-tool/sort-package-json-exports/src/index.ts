import { IPackageJson } from '@ts-type/package-dts/package-json';
import { sortObjectKeys } from 'sort-object-keys2';
import isPlainObject from 'is-plain-obj';
import {
	IPackageExportsEntryObject, IPackageExportsValue,
} from '@ts-type/package-dts/lib/package-json/exports';

export function isPackageJsonExportsEntryObject<T extends IPackageExportsValue, O extends IPackageExportsEntryObject = IPackageExportsEntryObject>(exports: T): exports is Extract<T, O>
{
	return isPlainObject(exports)
}

export interface IOptions
{
	rootOrder?: readonly string[];
	entryOrder?: readonly string[];
}

export function _handleOptions(options?: IOptions)
{
	const rootOrder = [
		'types',
		'require',
		'import',
		'node',
		'node-addons',
	];

	const entryOrder = [
		...rootOrder,
		'default',
		'.',
		'./',
		'./package.json',
	];

	if (options)
	{
		if (options.rootOrder?.length)
		{
			rootOrder.splice(0, 0, ...options.rootOrder);
		}
		if (options.entryOrder?.length)
		{
			entryOrder.splice(0, 0, ...options.entryOrder);
		}
	}

	return {
		rootOrder,
		entryOrder,
	}
}

export function sortPackageJsonExports(exports: IPackageJson["exports"], options?: IOptions)
{
	if (isPackageJsonExportsEntryObject(exports))
	{
		const { rootOrder, entryOrder } = _handleOptions(options);

		sortObjectKeys(exports, {
			keys: rootOrder,
			useSource: true,
		});

		Object.keys(exports)
			.forEach(key =>
			{
				const value = exports[key];

				if (isPackageJsonExportsEntryObject(value))
				{
					exports[key] = sortObjectKeys(value, {
						keys: entryOrder,
						useSource: true,
					});
				}
			})
	}

	return exports
}

export default sortPackageJsonExports
