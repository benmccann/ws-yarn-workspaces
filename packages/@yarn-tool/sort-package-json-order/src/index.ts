import { sortOrder } from './sort-order';

export { sortOrder }

export default sortOrder
