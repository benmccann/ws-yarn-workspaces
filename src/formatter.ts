import chalk from "chalk";
import Table from "cli-table";
import { Diff, DiffArray } from "deep-diff";

export const FormatterService = {
  _diffArray: (array: DiffArray<{}, {}>): [string, string] => {
    const item = array.item;
    switch (item.kind) {
      case "N":
        return [`[...]`, `[..., ${item.rhs}]`];
      case "D":
        return [`[..., ${item.lhs}]`, `[...]`];
      case "E":
        return [`[...], ${item.lhs}]`, `[..., ${item.rhs}]`];
      default:
        return [`[...]`, `[...]`];
    }
  },
  buildDiffTable: (diff: Diff<{}, {}>[]): string => {
    const table = new Table({
      head: [
        chalk.blueBright("package name"),
        chalk.blueBright("old version(s)"),
        chalk.blueBright("new version(s)")
      ],
      colWidths: [50, 20, 20]
    });
    diff.map(packageDiff => {
      switch (packageDiff.kind) {
        case "D":
          table.push([packageDiff.path, chalk.red(`${packageDiff.lhs}`), "-"]);
          break;
        case "N":
          table.push([
            packageDiff.path,
            "-",
            chalk.green(`${packageDiff.rhs}`)
          ]);
          break;
        case "E":
          table.push([
            packageDiff.path,
            chalk.yellow(`${packageDiff.lhs}`),
            chalk.yellow(`${packageDiff.rhs}`)
          ]);
          break;
        case "A":
          table.push([
            packageDiff.path,
            ...FormatterService._diffArray(packageDiff)
          ]);
      }
    });
    return table.toString();
  }
};
