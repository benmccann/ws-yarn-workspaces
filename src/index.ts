import { Command, flags } from "@oclif/command";
import * as fs from "fs";
import * as path from "path";

import { DiffService } from "./diff-service";
import { FormatterService } from "./formatter";

class YarnLockDiff extends Command {
  static description =
    "Given one or more old yarn.lock files and one or more new yarn.lock files, compute the diff";

  static flags = {
    version: flags.version({ char: "v" }),
    help: flags.help({ char: "h" }),
    old: flags.string({
      char: "o",
      description: "old yarn.lock file(s)",
      multiple: true,
      required: true
    }),
    new: flags.string({
      char: "n",
      description: "new yarn.lock file(s)",
      multiple: true,
      required: true
    })
  };

  checkIfExist = (filePathString: string) => {
    const filePath = path.resolve(process.cwd(), filePathString);
    return fs.existsSync(filePath);
  };

  reportNonExistantFiles = (filePathString: string) => {
    if (!this.checkIfExist(filePathString)) {
      this.error(`File does not exist "${filePathString}"`);
    }
    return filePathString;
  };

  checkIfLockFile = (filePathString: string) => {
    const fileExtension = path.extname(filePathString);
    return fileExtension === ".lock";
  };

  reportNonLockFile = (filePathString: string) => {
    if (!this.checkIfLockFile(filePathString)) {
      this.error(`File is not a lock file "${filePathString}"`);
    }
    return filePathString;
  };

  async run() {
    const { flags } = this.parse(YarnLockDiff);

    const oldYarnLockPaths = flags.old;
    const newYarnLockPaths = flags.new;

    [...oldYarnLockPaths, ...newYarnLockPaths]
      .map(this.reportNonExistantFiles)
      .map(this.reportNonLockFile);

    const oldContents = oldYarnLockPaths.map(path =>
      fs.readFileSync(path, "utf8")
    );
    const newContents = newYarnLockPaths.map(path =>
      fs.readFileSync(path, "utf8")
    );

    DiffService.buildDiff(oldContents, newContents)
      .map(FormatterService.buildDiffTable)
      .map(this.log);
  }
}

export = YarnLockDiff;
