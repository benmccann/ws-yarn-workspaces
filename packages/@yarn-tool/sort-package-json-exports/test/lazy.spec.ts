import { sortPackageJsonExports } from '../src/index';

test(`exports`, () =>
{
	let data = {
		"./k2": {

			// Entry-point for `require("my-package") in CJS
			"require": "./commonjs/index.cjs",
			// Entry-point for TypeScript resolution - must occur first!
			"types": "./types/index.d.ts",
			// Entry-point for `import "my-package"` in ESM
			"import": "./esm/index.js",
		},
		".": {
			// Entry-point for `import "my-package"` in ESM
			"import": "./esm/index.js",
			// Entry-point for TypeScript resolution - must occur first!
			"types": "./types/index.d.ts",
			// Entry-point for `require("my-package") in CJS
			"require": "./commonjs/index.cjs",
		},
	};

	let actual = sortPackageJsonExports(data);

	expect(actual).toMatchSnapshot();
	expect(JSON.stringify(actual, null, 2)).toMatchSnapshot();

});

