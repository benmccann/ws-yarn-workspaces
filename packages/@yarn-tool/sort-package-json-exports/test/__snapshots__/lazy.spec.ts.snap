// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exports 1`] = `
{
  ".": {
    "import": "./esm/index.js",
    "require": "./commonjs/index.cjs",
    "types": "./types/index.d.ts",
  },
  "./k2": {
    "import": "./esm/index.js",
    "require": "./commonjs/index.cjs",
    "types": "./types/index.d.ts",
  },
}
`;

exports[`exports 2`] = `
"{
  ".": {
    "types": "./types/index.d.ts",
    "require": "./commonjs/index.cjs",
    "import": "./esm/index.js"
  },
  "./k2": {
    "types": "./types/index.d.ts",
    "require": "./commonjs/index.cjs",
    "import": "./esm/index.js"
  }
}"
`;
